agent/lib/kontena/watchdog.rb

```diff
@@ -0,0 +1,71 @@
+# The watchdog is intended to recover the agent from unrepairable errors by killing it, and allowing it to be restarted.
+#
+# The watchdog takes a block, which gets run in a separate thread every interval.
+# The watchdog will abort if the block times out or fails with an exception.
+#
+# The watchdog abort will immediately exit! the process, without running any at_exit hooks for shutdown.
+class Kontena::Watchdog
+  include Celluloid
+  include Kontena::Logging
+
+  INTERVAL = 10.0
+  TIMEOUT = 30.0
+
+  def self.watch(**options, &block)
+    # pass block as a proc, instead of as a celluloid block proxy
+    new(block, **options)
+  end
+
+  def initialize(block, interval: INTERVAL, timeout: TIMEOUT, abort_exit: true)
+    @block = block
+
+    @interval = interval
+    @timeout = timeout
+    @abort_exit = abort_exit
+
+    async.start
+  end
+
+  def start
+    info "watchdog start"
+    @timer = every(@interval) do
+      ping
+    end
+  end
+
+  # Stop the every loop from start
+  def stop
+    @timer.cancel
+  end
+
+  # Start new watchdog ping
+  def ping
+    start = Time.now
+
+    # use exclusive mode to prevent the task from suspending and causing the Timeout::Error to crash the actor thread instead
+    exclusive {
+      Timeout.timeout(@timeout) do
```

Comment:
Consider replacing `Timeout` with something less dangerous.

Reply:
What part of `Timeout` being too dangerous do you mean, though?

The general problem with `Timeout.timeout` is that the non-`StandardError` exception (`Timeout::Error`) can get raised anywhere, and that might leave some code in an inconsistent state. However, that's not really a problem here: after a timeout, the watchdog will be `exit!`'ing the process anyways, so any inconsistent state / side-effects of the `Timeout::Error` shouldn't really matter anymore.

If you mean the interaction between celluloid task `Fiber` vs the `Timeout.timeout` aborting the celluloid actor `Thread`, then I could change this back to using a `defer` thread to run the block... but I don't think that's a problem with the `exclusive { ... }`? There's even spec cases for both kinds of suspends.

Reply:
It's not dangerous in this case. We can probably improve this later if Watchdog ever checks something more fragile.